sprotty-vscode-extension/src/webview-panel-manager.ts

```diff
@@ -0,0 +1,160 @@
+                this.endpointMap.set(key, endpoint);
+            }
+        }
+        return endpoint;
+    }
+
+    protected createEndpoint(identifier: SprottyDiagramIdentifier): WebviewEndpoint {
+        const webviewContainer = this.createWebview(identifier);
+        return new WebviewEndpoint({ webviewContainer, identifier });
+    }
+
+    /**
+     * Create and configure a webview panel. The default implementation sets `localResourceRoots` to the `pack` subfolder of the extension
+     * host and `scriptUri` to `pack/webview.js`. Please configure your bundler to generate such an output file from the Sprotty webview
+     * frontend. In case you need to use different settings or change the HTML content, you can override this functionality in a subclass.
+     */
+    protected createWebview(identifier: SprottyDiagramIdentifier): vscode.WebviewPanel {
+        const extensionPath = this.options.extensionUri.fsPath;
+        return createWebviewPanel(identifier, {
+            localResourceRoots: [ createFileUri(extensionPath, 'pack') ],
+            scriptUri: createFileUri(extensionPath, 'pack', 'webview.js')
+        });
+    }
+
+    protected async createDiagramIdentifier(uri: vscode.Uri, diagramType?: string): Promise<SprottyDiagramIdentifier | undefined> {
+        if (!diagramType) {
+            diagramType = await this.getDiagramType(uri);
+            if (!diagramType) {
+                return undefined;
+            }
+        }
+        const clientId = diagramType + '_' + WebviewPanelManager.viewCount++;
+        return {
+            diagramType,
+            uri: serializeUri(uri),
+            clientId
+        };
+    }
+
+    /**
+     * Determine a diagram type from the given URI. The default implementation returns the `defaultDiagramType` option
+     * if the URI matches the `supportedFileExtensions` or no file extensions were provided. If no default diagram type
+     * was provided, the file extension is used instead.
+     */
+    protected getDiagramType(uri: vscode.Uri): Promise<string | undefined> | string | undefined {
+        const extname = getExtname(uri);
+        if (!this.options.supportedFileExtensions || this.options.supportedFileExtensions.includes(extname)) {
+            if (this.options.defaultDiagramType) {
+                return this.options.defaultDiagramType;
+            }
+            if (extname.length >= 2) {
+                // Fallback: use the file extension after the '.' as diagram type
+                return extname.substring(1);
+            }
+        }
+    }
+
+    protected getKey(identifier: SprottyDiagramIdentifier) {
+        return `${identifier.diagramType}>>${identifier.uri}`;
+    }
+
+    protected didCloseWebview(endpoint: WebviewEndpoint): void {
+        if (this.singleton === endpoint) {
+            this.singleton = undefined;
+        } else {
+            const entry = Array.from(this.endpointMap.entries()).find(([key, value]) => value === endpoint);
```

Comment:
Can `this.getKey(endpoint.diagramIdentifier)` be used to get the key instead of iterating over the map entries?

Reply:
After looking at this again, I decided to keep a single array of endpoints instead of a map + singleton reference. @dhuebner could you review the change?